### DISCUSSION TITLE
Receive image using TcpClient ?

### DISCUSSION THREAD

**Discussion by seghier** [comfyanonymous/ComfyUI#3424] [Q&A]
Hello
I use Tcp to send image from c# to python, and everything works fine but i don't know how i can use python receiver in comfyui node?

c# sender
```
private void streamImage(Bitmap image)
{
    // Convert image to byte array
    byte[] imageData;
    using (MemoryStream stream = new MemoryStream())
    {
        image.Save(stream, System.Drawing.Imaging.ImageFormat.Jpeg);
        imageData = stream.ToArray();
    }

    // Connect to Python socket server
    using (TcpClient client = new TcpClient("127.0.0.1", 8888))
    {
        using (NetworkStream stream = client.GetStream())
        {
            // Send image data
            stream.Write(imageData, 0, imageData.Length);
        }
    }

    //Console.WriteLine("Image sent.");
}
```

python receiver
```
import socket
from PIL import Image
from io import BytesIO

def main():
    # Set up socket server
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.bind(('127.0.0.1', 8888))
    server.listen(1)

    print("Waiting for connection...")
    conn, addr = server.accept()
    print(f"Connected to {addr}")

    # Receive image data
    data = b""
    while True:
        chunk = conn.recv(1024)
        if not chunk:
            break
        data += chunk

    # Reconstruct image from byte array
    image = Image.open(BytesIO(data))

    # Display the received image
    image.show()

    print("Image received.")

if __name__ == "__main__":
    main()

```

ComfyUI node

```
class GHReceivedImage(GHNode):
    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "nickname": ("STRING", {"multiline": False}),
            }
        }
        
    RETURN_TYPES = ("IMAGE",)
    RETURN_NAMES = ("image",)
    FUNCTION = "load_image"
    OUTPUT_NODE = True
    CATEGORY = "Grasshopper"

    def __init__(self):
        self.server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.server.bind(('127.0.0.1', 8888))
        self.server.listen(1)
        print("Waiting for connection...")

    def load_image(self, nickname):
        try:
            conn, addr = self.server.accept()
            print(f"Connected to {addr}")

            data = b""
            while True:
                chunk = conn.recv(1024)
                if not chunk:
                    break
                data += chunk

            received_image = Image.open(BytesIO(data))
            return (received_image,)  # Return the image object wrapped in a tuple
        except Exception as e:
            print("Error occurred during image receiving:", e)
            return None
        finally:
            if 'conn' in locals():
                conn.close()  # Close the connection if it was established
```
i get this error

```
Traceback (most recent call last):
  File "E:\ComfyUI_windows_portable\ComfyUI\execution.py", line 151, in recursive_execute
    output_data, output_ui = get_output_data(obj, input_data_all)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\ComfyUI_windows_portable\ComfyUI\execution.py", line 81, in get_output_data
    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\ComfyUI_windows_portable\ComfyUI\execution.py", line 74, in map_node_over_list
    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\ComfyUI_windows_portable\ComfyUI\nodes.py", line 1404, in save_images
    full_output_folder, filename, counter, subfolder, filename_prefix = folder_paths.get_save_image_path(filename_prefix, self.output_dir, images[0].shape[1], images[0].shape[0])

                                    ~~~~~~^^^
TypeError: 'JpegImageFile' object is not subscriptable
```

**Comment by IARI** [comment]
Unfortunately I don't know how to help, but I am very very interested in achieving the same thing.
Let me know if you make any progress, or want to have a chat.

**Comment by seghier** [comment]
I solve it , but when i try to send a new image and queue selected node (preview image) , ComfyUI  show me this: got prompt
Prompt executed in 0.00 seconds and i can't receive a new image

```
class GHReceivedImage(GHNode):
    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "nickname": ("STRING", {"multiline": False}),
            }
        }
        
    RETURN_TYPES = ("IMAGE",)
    RETURN_NAMES = ("image",)
    FUNCTION = "load_image"
    OUTPUT_NODE = True
    CATEGORY = "Grasshopper"

    def __init__(self):
        super().__init__()  # Call superclass constructor
        self.server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.server.bind(('127.0.0.1', 8888))
        self.server.listen(1)
        print("Waiting for connection...")

    def receive_image(self):
        try:
            conn, addr = self.server.accept()
            print(f"Connected to {addr}")

            data = b""
            while True:
                chunk = conn.recv(1024)
                if not chunk:
                    break
                data += chunk

            return data
        except Exception as e:
            print("Error occurred during image receiving:", e)
            return None
        finally:
            if 'conn' in locals():
                conn.close()  # Close the connection if it was established

    def load_image(self, nickname):
        try:
            # Receive the image data
            image_data = self.receive_image()

            # If image data is received successfully
            if image_data:
                # Convert the image data to base64
                image_base64 = base64.b64encode(image_data).decode('utf-8')
                
                imgdata = base64.b64decode(image_base64)
                img = Image.open(BytesIO(imgdata))

                img = img.convert("RGB")
                img = np.array(img).astype(np.float32) / 255.0
                img = torch.from_numpy(img)[None,]
                
                # Return the base64-encoded image data
                return (img,)
            else:
                # If image data receiving fails, return None
                return None
        except Exception as e:
            print("Error occurred during image receiving:", e)
            return None
```

**Reply by TinyTerra** [reply]
You can use the IS_CHANGED method to force the node to re-execute even if it's inputs haven't changed. As seen in the [example_node.py.example](https://github.com/comfyanonymous/ComfyUI/blob/master/custom_nodes%2Fexample_node.py.example) file

**Comment by seghier** [comment]
Update, i add input for change the port and when i change it to any number it receive the new image
maybe i will add a way to random the port after closing the connection

**Comment by IARI** [comment]
One thing I wonder (for yours as well as my usecase):
Why do you have to create your own websocket connection?

Is it not an option to use the websocket connection that comfyui already establishes with its PromptServer?

**Reply by seghier** [reply]
Because i don't find a way to receive images directly from other softwares.
If comfyui websocket already can do that and you know how , i will use it

**Reply by IARI** [reply]
At least accessing the sockets from the server has been pretty easy (via PromptServer.instance.sockets )
So far I haven't been able to receive data yet, but I haven't tried very much, just found that it's not straightforward because aiohttp poses some constraints.

I am right now on vacation but will resume experimenting with it next week. 
I'll let you know, when I make progress.

**Reply by IARI** [reply]
I think reading messages from the socket could be a valuable comfy-ui feature for developers.
I can imagine there are more than just the two of us @seghier who want a third-party app to talk to comfy-ui via socket in custom nodes.

From what i understand, reading input from the socket has to be done in the same place where the aiohttp socket is initialised.

If that is correct, then there is no way around modifying the server.py in order to access messages from the regular websocket ( ws?clientId=... endpoint)
Right now messages are read in https://github.com/comfyanonymous/ComfyUI/blob/2de3b69b3074028fd0f850f801f4ca023489c692/server.py#L114-L116 but only errors are handled.

What I would like to have in the end is a method in the PromptServer, 
```python
def receive_message(self, sid, filter):
    pass
```
which allows for a node to wait for a specific socket message within its execution-function.
The filter argument should be an optional argument which can be a lambda that receives a message an the messages.
So in my custom node I could write something like
```python
# ...
FUNCTION = "receive_my_value_from_thirdparty_app"

def receive_my_value_from_thirdparty_app(self, name):
        server: PromptServer = PromptServer.instance
        sid: str = get_client_id() # get the client_id from the socket-connection of my third-party program
        myMessage = server.receive_message(sid, self.isValidMesage)
        return (self.process(myMessage),)
        
def process(self, msg):
        # process the message, i.e. make an image from a base64 string sent by the app.
        
def isValidMesage(self, msg):
        # check if the message is what I expected
```
I am trying to get that to work for me right now.

What do you think about that interface @seghier - would it be suitable for your use-case?

@comfyanonymous What do you think about this?

**Comment by seghier** [comment]
Thanks @IARI 
My tests before works fine and i add random seed option to make the node receive new image.
But i give up because that make a problem of many python processes and images are not saved.
It is a good way to send images in case of sketching or painting but it is not useful in my case with Rhino software.
But i still need a way to get the final image sent with websocket.

**Reply by IARI** [reply]
Hey @seghier ,

Not quite sure what you mean by

> It is a good way to send images in case of sketching or painting but it is not useful in my case with Rhino software.

Which way is a good way, but not useful with rhino?
 - the one that you tried?
 - or the interface that I proposed?
Or both?
I don't understand the problem you're facing

> But i still need a way to get the final image sent with websocket.

Do you mean the other direction, i.e. 
FROM comfy-ui output TO the client (c#/Rhino in your case) ?
Shouldn't that work easily with the existing nodes that save an image to the websocket?

**Comment by seghier** [comment]
As i said before i cancel it because it is not useful in my case
i use save images to a folder and use c# to get the latest image from output folder
This is the best method for now.
